perf(codegen): speed up printing `StringLiteral`s

overlookmotel opened the pull request:
Speed up printing `StringLiteral`s by:

* Iterating over the string byte-by-byte, not char-by-char.
* Not using a peekable iterator.
* Printing string in chunks.
* Using a lookup table for bytes that need escaping.
* Minimizing the amount of code in the hot loop.
* Calculating the quote character to use lazily.
* Faster calculation of quote character when backtick cannot be used as quote.

In common case where string contains no bytes that need escaping, the string will be written to output buffer in one shot, and there's no need for an extra pass through the string's bytes to decide on the quote character to use.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10046?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10046** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10046?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10046?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10044** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10044?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10041** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10041?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10042** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10042?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10040** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10040?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10039** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10039?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10036** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10036?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10046).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
I don't want you to go down the rabbit hole ... but we can probably avoid looping through the string twice by storing the position of the "unknown" values, and change them at the end of the loop.

Boshen commented:
### Merge activity

* **Mar 28, 12:29 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
> I don't want you to go down the rabbit hole ... but we can probably avoid looping through the string twice by storing the position of the "unknown" values, and change them at the end of the loop.

By "unknown" values, do you mean the quotes? We could do that, but it would be a bit of a pain because some characters need prefixing with `\` or not depending on what the quote character is, so "fixing up" these chars at the end would require shuffling up parts of the string to make room for the `\`s - it's not an "in place" modification.

Maybe that would be more optimal. But personally, I think the current solution is *okay*, even if it's not stellar. The main thing is that most strings don't contain any quote characters, `\n` or `$` - and for that common case, the 2nd pass through the string is avoided.

overlookmotel commented:
As a test, I tried the shortcut of just printing `raw` - the fastest possible way to print a `StringLiteral`, as it involves no processing at all. That only gave a 0.15% perf improvement vs this PR.

So basically, I think there's very little gain in trying to optimize it any further. We might get 0.1% perf improvement at best. It was probably a waste of time even going as far as I have in this PR!

overlookmotel commented:
This is *finally* ready to merge.

overlookmotel commented:
@Boshen I assume the effect on formatter benchmarks is just noise?

overlookmotel committed:
perf(codegen): speed up printing `StringLiteral`s (#10046)

Speed up printing `StringLiteral`s by:

* Iterating over the string byte-by-byte, not char-by-char.
* Not using a peekable iterator.
* Printing string in chunks.
* Using a lookup table for bytes that need escaping.
* Minimizing the amount of code in the hot loop.
* Calculating the quote character to use lazily.
* Faster calculation of quote character when backtick cannot be used as quote.

In common case where string contains no bytes that need escaping, the string will be written to output buffer in one shot, and there's no need for an extra pass through the string's bytes to decide on the quote character to use.
